src/cmd_all/src/single_node.rs

```diff
@@ -96,50 +139,282 @@ pub fn make_single_node_state_store_url(store_directory: &String) -> String {
+    let mut compactor_opts = RawOpts::default_compactor_opts();
+
     if let Some(prometheus_listener_addr) = &opts.prometheus_listener_addr {
-        meta_opts.prometheus_listener_addr = Some(prometheus_listener_addr.clone());
-        compute_opts.prometheus_listener_addr = prometheus_listener_addr.clone();
-        frontend_opts.prometheus_listener_addr = prometheus_listener_addr.clone();
-        compactor_opts.prometheus_listener_addr = prometheus_listener_addr.clone();
+        meta_opts.opts.insert(
+            "--prometheus-listener-addr".to_string(),
+            prometheus_listener_addr.clone(),
+        );
+        compute_opts.opts.insert(
+            "--prometheus-listener-addr".to_string(),
+            prometheus_listener_addr.clone(),
+        );
+        frontend_opts.opts.insert(
+            "--prometheus-listener-addr".to_string(),
+            prometheus_listener_addr.clone(),
+        );
+        compactor_opts.opts.insert(
+            "--prometheus-listener-addr".to_string(),
+            prometheus_listener_addr.clone(),
+        );
     }
     if let Some(config_path) = &opts.config_path {
-        meta_opts.config_path = config_path.clone();
-        compute_opts.config_path = config_path.clone();
-        frontend_opts.config_path = config_path.clone();
-        compactor_opts.config_path = config_path.clone();
+        meta_opts
+            .opts
+            .insert("--config-path".to_string(), config_path.clone());
+        compute_opts
+            .opts
+            .insert("--config-path".to_string(), config_path.clone());
+        frontend_opts
+            .opts
+            .insert("--config-path".to_string(), config_path.clone());
+        compactor_opts
+            .opts
+            .insert("--config-path".to_string(), config_path.clone());
     }
     if let Some(store_directory) = &opts.store_directory {
         let state_store_url = make_single_node_state_store_url(store_directory);
         let meta_store_endpoint = make_single_node_sql_endpoint(store_directory);
-        meta_opts.state_store = Some(state_store_url);
-        meta_opts.sql_endpoint = Some(meta_store_endpoint);
+        meta_opts
+            .opts
+            .insert("--state-store".to_string(), state_store_url);
+        meta_opts
+            .opts
+            .insert("--sql-endpoint".to_string(), meta_store_endpoint);
     }
     if let Some(meta_addr) = &opts.meta_addr {
-        meta_opts.listen_addr = meta_addr.clone();
-        meta_opts.advertise_addr = meta_addr.clone();
-
-        compute_opts.meta_address = meta_addr.parse().unwrap();
-        frontend_opts.meta_addr = meta_addr.parse().unwrap();
-        compactor_opts.meta_address = meta_addr.parse().unwrap();
+        meta_opts
+            .opts
+            .insert("--listen-addr".to_string(), meta_addr.to_string());
+        meta_opts
+            .opts
+            .insert("--advertise-addr".to_string(), meta_addr.to_string());
```

Comment:
A node's listen-addr can be different from its advertise-addr. This is common in K8s deployment, where nodes listen to local agent (`127.0.0.` or `0.0.0.0`) but other nodes need to reach it via network gateway (a domain name usually)

Reply:
These are `single-node`  process level opts, not meant for k8s deployment, so it is fine to override them like this.